lib/datadog/tracing/contrib/karafka/events/worker/process.rb

```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
+
+require_relative '../../ext'
+require_relative '../../event'
+
+module Datadog
+  module Tracing
+    module Contrib
+      module Karafka
+        module Events
+          module Worker
+            module Process
+              include Karafka::Event
+
+              def self.subscribe!
+                ::Karafka.monitor.subscribe 'worker.process' do |event|
+                  # Start a trace
+                  span = Tracing.trace(Ext::SPAN_WORKER_PROCESS, **span_options)
+
+                  job = event[:job]
+                  job_type = fetch_job_type(job.class)
+                  consumer = job.executor.topic.consumer
+                  topic = job.executor.topic.name
+
+                  action = case job_type
+                     when 'Periodic'
+                       'tick'
+                     when 'PeriodicNonBlocking'
+                       'tick'
+                     when 'Shutdown'
+                       'shutdown'
+                     when 'Revoked'
+                       'revoked'
+                     when 'RevokedNonBlocking'
+                       'revoked'
+                     when 'Idle'
+                       'idle'
+                      when 'Eofed'
+                       'eofed'
+                     when 'EofedNonBlocking'
+                       'eofed'
+                     else
+                       'consume'
+                     end
+
+                  span.resource = "#{consumer}##{action}"
+                  span.set_tag(Ext::TAG_TOPIC, topic) if topic
+
+                  if action == 'consume'
+                    span.set_tag(Ext::TAG_MESSAGE_COUNT, job.messages.count)
+                    span.set_tag(Ext::TAG_PARTITION, job.executor.partition)
+                    span.set_tag(Ext::TAG_OFFSET, job.messages.first.metadata.offset)
+                  end
+
+                  span
+                end
+
+                ::Karafka.monitor.subscribe 'worker.completed' do |event|
+                  Tracing.active_span&.finish
+                end
```

Comment:
Having separate locations for span creation `Tracing.trace` and span conclusion `Tracing.active_span&.finish` is always a possible source of hard-to-debug errors and span leakage (and thus memory leaks). We normally only do it when it's impossible to use `Tracing.trace { do_work_here }`.

Also, `Tracing.trace { do_work_here }` takes care of error handling, properly tagging the current span with error information.

In this case, we have the event `worker.processed` that looks like it's just what we need:
https://github.com/karafka/karafka/blob/ab4f9bcd3620f46adb8c0d158b5396b245619ed3/lib/karafka/processing/worker.rb#L58-L78

_**Except**_ that it doesn't call the event listeners when an error is raised by the job. There are no error handlers in this method that call `assigned_listeners`: https://github.com/karafka/karafka-core/blob/a1425725d275796673424c1cd9be517d06518ec9/lib/karafka/core/monitoring/notifications.rb#L120

I opened a PR to Karafka to address this, but even if it's approved, it won't affect users of older versions of the library: https://github.com/karafka/karafka-core/pull/145

The being said, I still lean towards using a single `Tracing.trace { do_work_here }` only because the safety of not having to worry about leaky spans is too advantageous.

Reply:
That’s precisely what I’d like to hear from the Datadog team regarding this. My initial intuition when using `Tracing.active_span` was that I couldn’t be certain whether the span I’m currently working on is the one I want to complete. Considering this and your insights, would it be feasible to remove this tracing event? At the moment, I prefer to retain the `Ext::SPAN_MESSAGE_CONSUME` event only for this integration, as it effectively meets my requirement: enabling distributed tracing, ensuring that each message span is linked to the root trace in the distributed tracing system.

Reply:
> I opened a PR to Karafka to address this, but even if it's approved, it won't affect users of older versions of the library

This PR is not needed to achieve the expected wrapping because it is already done for example for OpenTelemetry:

ref1: https://karafka.io/docs/Monitoring-and-Logging/#opentelemetry
ref2: https://karafka.io/docs/Monitoring-and-Logging/#monitor-wrapping-and-replacement